### PR TITLE
GraalVM support

### PR DESCRIPTION
**Pull request by msailes**
*Issue #, if available:*

*Description of changes:*
GraalVM config and regression test.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by szab100** [comment]
Hi @msailes, what is this PR waiting for? We are facing with the same issues when trying to use spring native with graalvm..

**Comment by msailes** [comment]
@szab100 if possible could you contact me at msailes@amazon.co.uk

Thanks,

**Comment by msailes** [comment]
@szab100 could you 👍  this issue please. https://github.com/aws/aws-lambda-java-libs/issues/272

**Comment by stejsj** [comment]
@msailes should the runtime client work with graalvm? We use scala and are trying to use your approach to get a native-image running to reduce cold start and resource requirements. 

Sadly I get this error 
START RequestId: 3a579869-9b96-46d7-9ce4-8c049a79d6c1 Version: $LATEST
[appconfig agent] 2022/04/05 10:36:21 INFO AppConfig Lambda Extension 2.0.45
[appconfig agent] 2022/04/05 10:36:21 INFO serving on localhost:2772
starting outage runtime...
Failed to load the native runtime interface client library aws-lambda-runtime-interface-client.glibc.so. Exception: null
Failed to load the native runtime interface client library aws-lambda-runtime-interface-client.musl.so. Exception: null
EXTENSION	Name: AppConfigAgent	State: Ready	Events: [SHUTDOWN,INVOKE]
END RequestId: 3a579869-9b96-46d7-9ce4-8c049a79d6c1
REPORT RequestId: 3a579869-9b96-46d7-9ce4-8c049a79d6c1	Duration: 253.23 ms	Billed Duration: 254 ms	Memory Size: 2048 MB	Max Memory Used: 56 MB	
XRAY TraceId: 1-624c1ba4-441854bd0da7b1425ca04379	SegmentId: 14184b9f35ddfe28	Sampled: true	

Eventhough I tried to manually include those two libraries as native-image resource args
example -H:IncludeResources="+baseDirectory.value / "src" / "main" / "resources" / "META-INF" / "resources" / "aws-lambda-runtime-interface-client.glibc.so"

We have been struggling quite a lot getting this to work, any thoughts, hints or ideas?

Thanx in advance
Steffen

File: aws-lambda-java-events/src/main/java/com/amazonaws/services/lambda/runtime/graalvm/LambdaEventsGraalVMFeature.java [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+ * native binary it creates.
+ *
+ * Any event or response class added to this library needs to be added to this config file.
+ *
+ * This implementation of Feature does that by iterating through the events package and registering
+ * them for runtime reflection.
+ */
+public class LambdaEventsGraalVMFeature implements Feature {
+
+    public static final String EVENTS_PACKAGE_NAME = "com.amazonaws.services.lambda.runtime.events";
```
Comment (carlzogh) [inline aws-lambda-java-events/src/main/java/com/amazonaws/services/lambda/runtime/graalvm/LambdaEventsGraalVMFeature.java:30]: I think we should make this generic enough for it to be reused across all our libraries in this project, and move it out of the runtime addressable scope. It should be a build-only class that is not exposed in the library.

This is essentially boilerplate code that will traverse our libraries and determine what the appropriate `native-image` config is, so we don't really need a copy of it per project.

File: aws-lambda-java-runtime-interface-client/src/main/resources/META-INF/native-image/com.amazonaws/aws-lambda-java-runtime-interface-client/jni-config.json [hunk 2]
```diff
@@ -0,0 +1,11 @@
...
+{
+  "name":"com.amazonaws.services.lambda.runtime.api.client.runtimeapi.LambdaRuntimeClientException",
+  "methods":[{"name":"<init>","parameterTypes":["java.lang.String","int"] }]
+},
+{
+  "name":"com.amazonaws.services.lambda.runtime.api.client.runtimeapi.InvocationRequest",
+  "fields":[{"name":"id"}, {"name":"invokedFunctionArn"}, {"name":"deadlineTimeInMs"}, {"name":"xrayTraceId"}, {"name":"clientContext"}, {"name":"cognitoIdentity"}, {"name":"content"}],
+  "allPublicMethods":true
+}
+]
```
Comment (carlzogh) [inline aws-lambda-java-runtime-interface-client/src/main/resources/META-INF/native-image/com.amazonaws/aws-lambda-java-runtime-interface-client/jni-config.json:11]: Can this be auto-generated like `reflect-config.json`?

File: aws-lambda-java-runtime-interface-client/test/integration/codebuild-graalvm/buildspec.os.amazoncorretto.yml [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+          -f "${SCRATCH_DIR}/Dockerfile.function.graalvm.tmp" \
+          -t "${IMAGE_TAG}" \
+          --build-arg RUNTIME_VERSION="${RUNTIME_VERSION}" \
+          --build-arg DISTRO_VERSION="${DISTRO_VERSION}"
+  build:
+    commands:
+      - set -x
+      - echo "Running Image ${IMAGE_TAG}"
+      - docker network create "${OS_DISTRIBUTION}-network"
+      - >
```
Comment (carlzogh) [inline aws-lambda-java-runtime-interface-client/test/integration/codebuild-graalvm/buildspec.os.amazoncorretto.yml:64]: Some parts of this script have been refactored, we should update this

**Review by carlzogh — CHANGES_REQUESTED** [review]

**Pull request closed without merging**